Source/Core/Common/Debug/CodeTrace.h

```diff
@@ -0,0 +1,70 @@
+// Copyright 2022 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <optional>
+#include <set>
+#include <string>
+#include <vector>
+#include "Common/CommonTypes.h"
+
+struct InstructionAttributes
+{
+  u32 address = 0;
+  std::string instruction = "";
+  std::string reg0 = "";
+  std::string reg1 = "";
+  std::string reg2 = "";
+  std::string reg3 = "";
+  std::optional<u32> memory_target = std::nullopt;
+  u32 memory_target_size = 4;
+  bool is_store = false;
+  bool is_load = false;
+};
+
+struct TraceOutput
+{
+  u32 address;
+  std::optional<u32> memory_target = std::nullopt;
+  std::string instruction;
+};
+
+enum class HitType : u32
+{
+  SKIP = (1 << 0),       // Not a hit
+  OVERWRITE = (1 << 1),  // Tracked value gets overwritten by untracked. Typically skipped.
+  MOVED = (1 << 2),      // Target duplicated to another register, unchanged.
+  SAVELOAD = (1 << 3),   // Target saved or loaded. Priority over Pointer.
+  POINTER = (1 << 4),    // Target used as pointer/offset for save or load
+  PASSIVE = (1 << 5),    // Conditional, etc, but not pointer. Unchanged
+  ACTIVE = (1 << 6),     // Math, etc. Changed.
+  UPDATED = (1 << 7)     // Masked or math without changing register.
```

Comment:
It's recommended that the last element also has a comma to avoid noise in future diffs